# TesnorFlow Problem Sheet Solution:

This program uses machine learning through tensor flow to identify flowers based on the IRIS dataset. First we import the dataset using pandas read_csv without headers because we already have our own ones made within the csv. We then create the feature columns for the model and set the shape as 4 since there are 4 columns our model will use. A DNNClassifier Model is created for comparing our dataset to and testing its accuacy. We then print the whole dataset to the user before using pandas to split the dataset in half. Once the dataset is split we save the training and testing sets to there own csvs. We then use the csvs to create seperate datasets and input. After testing these we compare the test to our model and print the accuracy. It should print out a result between 0.9-1.0. Then we use the test set to predict the type of flowers in the data.

## Imports:
- Numpy: We import this library for complex math equations and for multi-dimensional arrays and matrices that the library allows us to use, these will be useful when manipulating our data.
- Tensorflow: Import tensorflow for machine learning algorithms.
- Pandas: This is a library for mainpulating csvs and is what we will use to seperate the data. 


In [1]:
import os
import numpy as np
import tensorflow as tf
import pandas as pd

## Use Tensorflow to create model
Here we identify our CSV containing our data, this is the IRIS dataset which is what we will be using later for training and testing. We import this into our program using pandas by passing it the CSV name and telling pandas not to assign headers as we all ready have them in the CSV. We then create our feature columns, these are used to describe to the Model what tensorflow will be learning and the dataset it will be using. We then create the model which will be used to assess the dataset. The model we use here is known as a Deep Neural Network Model or DNN, this is used to establish correlations between data which makes it ideal for the IRIS dataset since we will training to establish the correlation and idetifing species from a test set using our resulting correlations from the training.

In [3]:
# IRIS data file name set up
IRIS = "IRIS.csv"

# Load data from csv.
print("Loading IRIS..")
fulldata = pd.read_csv(IRIS, header=None)
  
# Prepare data for modeling by checking the columns with tf.feature_column.numeric_column.
# We have the Petal length + width and the Speal length + width so we set the shape to 4 since there are 4 columns to feature
feature_columns = [tf.feature_column.numeric_column("x", shape=[4])]

# Fit our data to a Deep Neural Network Classifier Model and make a temp directory to hold it
print("Creating Model..")
classifier = tf.estimator.DNNClassifier(feature_columns=feature_columns,
                                          hidden_units=[10, 20, 10],
                                          n_classes=3,
                                          model_dir="Model")

Loading IRIS..
Creating Model..
INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_model_dir': 'Model', '_tf_random_seed': 1, '_save_summary_steps': 100, '_save_checkpoints_secs': 600, '_save_checkpoints_steps': None, '_session_config': None, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100}


## Split the data into training and testing

First thing I do here is output the whole dataset to the screen so the user can see all the flower species, this isnt necessary but allows users to go back and compare the tensorflow results to the original dataset. I then seperate the dataset into training and testing, I use pandast to do this by telling training to take it first 100 rows of the IRIS dataset and testing to take the last 50. I stop pandas from setting index headers and then apply my own headers using the names parameter.The names format is: rows - columns - setosa - versicolor - virginica as this is how the format the DNN will use to read the data later. I then save each set to a csv from pandas without indexs and includeding the headers ive set with names. Now that the data is neatly seperated into CSVs ill use them to set up tensorflow datasets by loading them in.

In [5]:
# Display the full dataset to the user and
# Split our mixed up IRIS data into training and testing
print("IRIS Dataset:")
print(fulldata)
training_set = pd.read_csv(IRIS, header=None, nrows=100 , names = ["100", "4", "setosa", "versicolor", "virginica"])
test_set = pd.read_csv(IRIS, header=None, skiprows=100, nrows=151, names = ["51", "4", "setosa", "versicolor", "virginica"])

# Set the file name for both CSV
train_data = "trainIRIS.csv"
test_data = "testIRIS.csv"

# Save them to individual CSV files
training_set.to_csv(train_data, index = False, header=True)
test_set.to_csv(test_data, index = False, header=True)

# Create training dataset
training_set = tf.contrib.learn.datasets.base.load_csv_with_header(
      filename=train_data,
      target_dtype=np.int,
      features_dtype=np.float32)

# Create testing dataset
test = tf.contrib.learn.datasets.base.load_csv_with_header(
      filename=test_data,
      target_dtype=np.int,
      features_dtype=np.float32)

IRIS Dataset:
       0    1    2    3  4
0    6.4  2.8  5.6  2.2  2
1    5.0  2.3  3.3  1.0  1
2    4.9  2.5  4.5  1.7  2
3    4.9  3.1  1.5  0.1  0
4    5.7  3.8  1.7  0.3  0
5    4.4  3.2  1.3  0.2  0
6    5.4  3.4  1.5  0.4  0
7    6.9  3.1  5.1  2.3  2
8    6.7  3.1  4.4  1.4  1
9    5.1  3.7  1.5  0.4  0
10   5.2  2.7  3.9  1.4  1
11   6.9  3.1  4.9  1.5  1
12   5.8  4.0  1.2  0.2  0
13   5.4  3.9  1.7  0.4  0
14   7.7  3.8  6.7  2.2  2
15   6.3  3.3  4.7  1.6  1
16   6.8  3.2  5.9  2.3  2
17   7.6  3.0  6.6  2.1  2
18   6.4  3.2  5.3  2.3  2
19   5.7  4.4  1.5  0.4  0
20   6.7  3.3  5.7  2.1  2
21   6.4  2.8  5.6  2.1  2
22   5.4  3.9  1.3  0.4  0
23   6.1  2.6  5.6  1.4  2
24   7.2  3.0  5.8  1.6  2
25   5.2  3.5  1.5  0.2  0
26   5.8  2.6  4.0  1.2  1
27   5.9  3.0  5.1  1.8  2
28   5.4  3.0  4.5  1.5  1
29   6.7  3.0  5.0  1.7  1
..   ...  ...  ...  ... ..
120  5.9  3.0  4.2  1.5  1
121  6.9  3.1  5.4  2.1  2
122  5.1  3.3  1.7  0.5  0
123  6.0  3.4  4.5  1.6  1
124  5.5  2.5 

## Train the model

Below I set up the inputs we will be using to teach our model how to identify the flower species. I use the training set data I set up for this. Then I call the DNN model to start training using the input, It will analyse the input 2000 times for correlations. 

In [6]:
# Define the training inputs
print("Preparing training with inputs..")
train_input_fn = tf.estimator.inputs.numpy_input_fn(
      x={"x": np.array(training_set.data)},
      y=np.array(training_set.target),
      num_epochs=None,
      shuffle=True)

# Train model.
print("Training model..")
classifier.train(input_fn=train_input_fn, steps=2000)


Preparing training with inputs..
Training model..
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Restoring parameters from Model\model.ckpt-162001
INFO:tensorflow:Saving checkpoints for 162002 into Model\model.ckpt.
INFO:tensorflow:loss = 0.260926, step = 162002
INFO:tensorflow:global_step/sec: 685.979
INFO:tensorflow:loss = 0.374497, step = 162102 (0.114 sec)
INFO:tensorflow:global_step/sec: 958.854
INFO:tensorflow:loss = 0.289361, step = 162202 (0.106 sec)
INFO:tensorflow:global_step/sec: 1241.24
INFO:tensorflow:loss = 0.435002, step = 162302 (0.094 sec)
INFO:tensorflow:global_step/sec: 862.951
INFO:tensorflow:loss = 0.285164, step = 162402 (0.116 sec)
INFO:tensorflow:global_step/sec: 672.287
INFO:tensorflow:loss = 0.523058, step = 162502 (0.149 sec)
INFO:tensorflow:global_step/sec: 549.617
INFO:tensorflow:loss = 0.335817, step = 162602 (0.171 sec)
INFO:tensorflow:global_step/sec: 760.131
INFO:tensorflow:loss = 0.160661, step = 162702 (0.130 sec)
INFO:tensorflow:global_s

## Test the model

Here we set up more inputs except this time we use the test set. We then pass these inputs to the DNN Model to test its accuacy. This should give back a accuracy higher the 0.90 if everything is working correctly. After we know how accurate out model is we can now use it for predictions. We take the same test input and pass it to our model so it can try estimate what are the species that each row of data describes. We convert the object holding the results into a list so it can be easily displayed. The reults are then printed to the screen by mapping each value so its easier for the user to read.

In [7]:
# Define the test inputs
test_input_fn = tf.estimator.inputs.numpy_input_fn(
      x={"x": np.array(test.data)},
      y=np.array(test.target),
      num_epochs=1,
      shuffle=False)

# Evaluate accuracy using test input on classifer model.
accuracy_score = classifier.evaluate(input_fn=test_input_fn)["accuracy"]

# Print model accuracy
print("\nTest Accuracy: {0:f}\n".format(accuracy_score))

# Predict the test array
print("Predicting data..")
predict_input_fn = tf.estimator.inputs.numpy_input_fn(x={"x": np.array(test.data)},
y=np.array(test.target),
num_epochs=1,
shuffle=False)

# Get predictied classes results from output
predictions = list(classifier.predict(input_fn=predict_input_fn))
predicted_classes = [p["classes"] for p in predictions]

# Display predicted classes
print("Prediction results:")
print('[%s]' % ', '.join(map(str, predicted_classes)))



INFO:tensorflow:Starting evaluation at 2017-11-21-18:14:13
INFO:tensorflow:Restoring parameters from Model\model.ckpt-164001
INFO:tensorflow:Finished evaluation at 2017-11-21-18:14:13
INFO:tensorflow:Saving dict for global step 164001: accuracy = 0.980392, average_loss = 0.144435, global_step = 164001, loss = 7.3662

Test Accuracy: 0.980392

Predicting data..
INFO:tensorflow:Restoring parameters from Model\model.ckpt-164001
Prediction results:
[[b'0'], [b'1'], [b'0'], [b'1'], [b'0'], [b'0'], [b'0'], [b'0'], [b'1'], [b'0'], [b'2'], [b'1'], [b'0'], [b'2'], [b'0'], [b'1'], [b'1'], [b'0'], [b'0'], [b'1'], [b'1'], [b'2'], [b'0'], [b'1'], [b'1'], [b'1'], [b'0'], [b'1'], [b'1'], [b'2'], [b'2'], [b'0'], [b'2'], [b'1'], [b'1'], [b'0'], [b'1'], [b'0'], [b'0'], [b'2'], [b'0'], [b'1'], [b'2'], [b'1'], [b'1'], [b'1'], [b'0'], [b'1'], [b'2'], [b'1'], [b'0']]


### Conclusion

The program gave me a 0.980392 accuracy and all the printed pridictions seem to be correct as far as I can see, this problem sheet gave me a good understanding of how to use tensorflow practically which will be helpfull in my project and drastically improved my understanding of Deep Neural Networks in general. I included a normal .py file in the repsoitory for users without jupyter notebook as well.